### PR TITLE
Add Fused CrossEntropy and Update Vocab Sizes

### PR DESCRIPTION
**Pull request by nik-mosaic**
Add Fused Cross Entropy, an (almost) math-equivalent loss function to torch.nn.CrossEntropy.

This loss function's requirements are installed via new requirements in llm/requirements.txt. We default to turning this on, and will throw an error if you are on a CPU and have not turned it off. You can turn it off with a new config `model.loss_fn = torch_crossentropy` and turn it on omitting loss_fn or setting `model.loss_fn = fused_crossentropy`.

This PR also updates the vocab sizes to be a multiple of 64, accounting for the new tokenizer.

### CODE REVIEW

File: examples/llm/src/models/mosaic_gpt.py [hunk 1]
```diff
@@ -308,6 +308,19 @@ def __init__(self, om_model_config: DictConfig):
             'Perplexity':
                 Perplexity(),
         }
+        if self.hf_config.loss_fn == 'fused_crossentropy':
+            try:
+                from flash_attn.losses.cross_entropy import CrossEntropyLoss as FusedCrossEntropyLoss  # type: ignore # isort: skip
+                print('Using Fused Cross Entropy Loss.')
+                self.loss_fn = FusedCrossEntropyLoss(inplace_backward=False,
```
Comment (vchiley) [inline examples/llm/src/models/mosaic_gpt.py:317]: why `inplace_backward=False`?
(maybe more for my curiosity, less for the PR; I'd expect `True` to use less mem, y set false?)

if `inplace_backward=False` is the default, should we be setting it again?

File: examples/llm/src/models/mosaic_gpt.py [hunk 2]
```diff
@@ -308,6 +308,19 @@ def __init__(self, om_model_config: DictConfig):
...
+                                                     ignore_index=-100,
+                                                     process_group=None)
```
Comment (vchiley) [inline examples/llm/src/models/mosaic_gpt.py:319]: does `process_group` default to None? if so, rm setting it.

File: examples/llm/requirements.txt [hunk 3]
```diff
@@ -11,3 +11,6 @@ omegaconf==2.2.3
 wandb==0.13.6
 pytest>=7.2.1,<8
 torchmetrics==0.11.3
+xentropy-cuda-lib@git+https://github.com/HazyResearch/flash-attention.git@v0.2.8#subdirectory=csrc/xentropy
+build==0.10.0
+ninja==1.11.1
```
Comment (vchiley) [inline examples/llm/requirements.txt:16]: r these needed for the xentropy install? or were they needed for the dropoutresiduallayernorm stuff?

File: examples/llm/src/models/mosaic_gpt.py [hunk 4]
```diff
@@ -308,6 +308,19 @@ def __init__(self, om_model_config: DictConfig):
...
+            except:
+                raise ValueError(
+                    'Fused Cross Entropy is not installed. Either (1) have a CUDA-compatible GPU and `pip install .[llm]`, or (2) set your config model.loss_fn=torch_crossentropy.'
+                )
+        else:
```
Comment (vchiley) [inline examples/llm/src/models/mosaic_gpt.py]: Isn't it more correct to do
`elif self.hf_config.loss_fn == 'torch_crossentropy'`
then `else: raise ValueError(f'specified loss_fn {self.hf_config.loss_fn} not recognized')`

File: examples/llm/src/models/mosaic_gpt.py [hunk 5]
```diff
@@ -308,6 +308,25 @@ def __init__(self, om_model_config: DictConfig):
             'Perplexity':
                 Perplexity(),
         }
+        print(resolved_om_config)
```
Comment (vchiley) [inline examples/llm/src/models/mosaic_gpt.py:311]:
```suggestion
```

File: examples/llm/yamls/mosaic_gpt/testing.yaml [hunk 6]
```diff
@@ -17,12 +17,13 @@ model:
...
   mlp_ratio: 4
   max_seq_len: ${max_seq_len}
-  vocab_size: 50257
+  vocab_size: 50368
   init_std: 0.02
   attn_pdrop: 0.0
   resid_pdrop: 0.0
   emb_pdrop: 0.0
   attn_impl: torch
+  loss_fn: torch_crossentropy
```
Comment (vchiley) [inline examples/llm/yamls/mosaic_gpt/testing.yaml:26]: added in two places?

**Review by vchiley — COMMENTED** [review]
Left a few comments.

File: examples/llm/tests/test_model.py [hunk 7]
```diff
@@ -321,6 +321,59 @@ def test_determinism(attention_type: str, precision):
...
+        loss_1 = model_1.loss(output_1, batch)
+        loss_2 = model_2.loss(output_2, batch)
+        assert loss_1.allclose(loss_2, rtol=1e-3,
+                               atol=1e-3), f'differed at step {i}'
+        loss_1.backward()
+        loss_2.backward()
+        optimizer_1.step()
+        optimizer_2.step()
+
+        for p1, p2 in zip(model_1.parameters(), model_2.parameters()):
```
Comment (vchiley) [inline examples/llm/tests/test_model.py:371]: is it **_guaranteed_** that the paramters() generator will generate parameters in the same order?
fearing the answer is no, I did [this here](https://github.com/mosaicml/examples/blob/952f4fe8a990b0540d6a989bd188f75ce67fa2f5/examples/llm/tests/test_flash_triton_torch.py#L106)

File: examples/llm/tests/test_model.py [hunk 8]
```diff
@@ -321,6 +321,59 @@ def test_determinism(attention_type: str, precision):
             optimizer_2.step()
 
 
+@pytest.mark.gpu
+def test_loss_fn():
```
Comment (vchiley) [inline examples/llm/tests/test_model.py:325]: add a blurb about what is being tested.

File: examples/llm/src/models/mosaic_gpt.py [hunk 9]
```diff
@@ -308,6 +308,22 @@ def __init__(self, om_model_config: DictConfig):
             'Perplexity':
                 Perplexity(),
         }
+        self.loss_fn = resolved_om_config.get('loss_fn', 'fused_crossentropy')
+        if self.loss_fn == 'fused_crossentropy':
+            try:
+                from flash_attn.losses.cross_entropy import CrossEntropyLoss as FusedCrossEntropyLoss  # type: ignore # isort: skip
+                print('Using Fused Cross Entropy Loss.')
```
Comment (vchiley) [inline examples/llm/src/models/mosaic_gpt.py:315]: warning.warn instead of print?

**Review by abhi-mosaic — APPROVED** [review]
Looks good and excited to start using this in all our LLM runs, thank you @nik-mosaic !

**Review by vchiley — APPROVED** [review]
